# **STEP 1: IMPORT LIBRARIES**

In [33]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# **STEP 2: LOAD DATA**

In [34]:
df = pd.read_csv("/content/credit_score.csv")
df

/tmp/ipython-input-1294880488.py:1: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/content/credit_score.csv")


ID Customer_ID     Month           Name   Age          SSN  \
0       0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265   
1       0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265   
2       0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265   
3       0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265   
4       0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265   
...        ...         ...       ...            ...   ...          ...   
99995  0x25fe9  CUS_0x942c     April          Nicks    25  078-73-5990   
99996  0x25fea  CUS_0x942c       May          Nicks    25  078-73-5990   
99997  0x25feb  CUS_0x942c      June          Nicks    25  078-73-5990   
99998  0x25fec  CUS_0x942c      July          Nicks    25  078-73-5990   
99999  0x25fed  CUS_0x942c    August          Nicks    25  078-73-5990   

      Occupation Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  ...  \
0      Scientist      19114.12            1824.843333                  3  ...   
1      Scientist      19114.12                    NaN                  3  ...   
2      Scientist      19114.12                    NaN                  3  ...   
3      Scientist      19114.12                    NaN                  3  ...   
4      Scientist      19114.12            1824.843333                  3  ...   
...          ...           ...                    ...                ...  ...   
99995   Mechanic      39628.99            3359.415833                  4  ...   
99996   Mechanic      39628.99            3359.415833                  4  ...   
99997   Mechanic      39628.99            3359.415833                  4  ...   
99998   Mechanic      39628.99            3359.415833                  4  ...   
99999   Mechanic     39628.99_            3359.415833                  4  ...   

       Credit_Mix  Outstanding_Debt Credit_Utilization_Ratio  \
0               _            809.98                26.822620   
1            Good            809.98                31.944960   
2            Good            809.98                28.609352   
3            Good            809.98                31.377862   
4            Good            809.98                24.797347   
...           ...               ...                      ...   
99995           _            502.38                34.663572   
99996           _            502.38                40.565631   
99997        Good            502.38                41.255522   
99998        Good            502.38                33.638208   
99999        Good            502.38                34.192463   

           Credit_History_Age  Payment_of_Min_Amount Total_EMI_per_month  \
0       22 Years and 1 Months                     No           49.574949   
1                         NaN                     No           49.574949   
2       22 Years and 3 Months                     No           49.574949   
3       22 Years and 4 Months                     No           49.574949   
4       22 Years and 5 Months                     No           49.574949   
...                       ...                    ...                 ...   
99995   31 Years and 6 Months                     No           35.104023   
99996   31 Years and 7 Months                     No           35.104023   
99997   31 Years and 8 Months                     No           35.104023   
99998   31 Years and 9 Months                     No           35.104023   
99999  31 Years and 10 Months                     No           35.104023   

      Amount_invested_monthly                 Payment_Behaviour  \
0           80.41529543900253   High_spent_Small_value_payments   
1          118.28022162236736    Low_spent_Large_value_payments   
2             81.699521264648   Low_spent_Medium_value_payments   
3           199.4580743910713    Low_spent_Small_value_payments   
4          41.420153086217326  High_spent_Medium_value_payments   
...                       ...                               ...   

# **STEP 3: DATA CLEANING**

In [35]:
df.isnull().sum()

ID                              0
Customer_ID                     0
Month                           0
Name                         9985
Age                             0
SSN                             0
Occupation                      0
Annual_Income                   0
Monthly_Inhand_Salary       15002
Num_Bank_Accounts               0
Num_Credit_Card                 0
Interest_Rate                   0
Num_of_Loan                     0
Type_of_Loan                11408
Delay_from_due_date             0
Num_of_Delayed_Payment       7002
Changed_Credit_Limit            0
Num_Credit_Inquiries         1965
Credit_Mix                      0
Outstanding_Debt                0
Credit_Utilization_Ratio        0
Credit_History_Age           9030
Payment_of_Min_Amount           0
Total_EMI_per_month             0
Amount_invested_monthly      4479
Payment_Behaviour               0
Monthly_Balance              1200
Credit_Score                    0
dtype: int64

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

In [37]:
df.describe()

Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
count           84998.000000      100000.000000     100000.00000   
mean             4194.170850          17.091280         22.47443   
std              3183.686167         117.404834        129.05741   
min               303.645417          -1.000000          0.00000   
25%              1625.568229           3.000000          4.00000   
50%              3093.745000           6.000000          5.00000   
75%              5957.448333           7.000000          7.00000   
max             15204.633333        1798.000000       1499.00000   

       Interest_Rate  Delay_from_due_date  Num_Credit_Inquiries  \
count  100000.000000        100000.000000          98035.000000   
mean       72.466040            21.068780             27.754251   
std       466.422621            14.860104            193.177339   
min         1.000000            -5.000000              0.000000   
25%         8.000000            10.000000              3.000000   
50%        13.000000            18.000000              6.000000   
75%        20.000000            28.000000              9.000000   
max      5797.000000            67.000000           2597.000000   

       Credit_Utilization_Ratio  Total_EMI_per_month  
count             100000.000000        100000.000000  
mean                  32.285173          1403.118217  
std                    5.116875          8306.041270  
min                   20.000000             0.000000  
25%                   28.052567            30.306660  
50%                   32.305784            69.249473  
75%                   36.496663           161.224249  
max                   50.000000         82331.000000

In [38]:
# Remove unwanted columns if exist
drop_cols = ['ID','Customer_ID','Name','SSN','Type_of_Loan']

In [39]:
df.drop(columns=drop_cols, errors='ignore', inplace=True)

In [40]:
df['Occupation'].unique()

array(['Scientist', '_______', 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Lawyer', 'Media_Manager', 'Doctor', 'Journalist',
       'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer',
       'Architect'], dtype=object)

In [41]:
df.head()

Month   Age Occupation Annual_Income  Monthly_Inhand_Salary  \
0   January    23  Scientist      19114.12            1824.843333   
1  February    23  Scientist      19114.12                    NaN   
2     March  -500  Scientist      19114.12                    NaN   
3     April    23  Scientist      19114.12                    NaN   
4       May    23  Scientist      19114.12            1824.843333   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate Num_of_Loan  \
0                  3                4              3           4   
1                  3                4              3           4   
2                  3                4              3           4   
3                  3                4              3           4   
4                  3                4              3           4   

   Delay_from_due_date  ... Credit_Mix Outstanding_Debt  \
0                    3  ...          _           809.98   
1                   -1  ...       Good           809.98   
2                    3  ...       Good           809.98   
3                    5  ...       Good           809.98   
4                    6  ...       Good           809.98   

   Credit_Utilization_Ratio     Credit_History_Age Payment_of_Min_Amount  \
0                 26.822620  22 Years and 1 Months                    No   
1                 31.944960                    NaN                    No   
2                 28.609352  22 Years and 3 Months                    No   
3                 31.377862  22 Years and 4 Months                    No   
4                 24.797347  22 Years and 5 Months                    No   

   Total_EMI_per_month Amount_invested_monthly  \
0            49.574949       80.41529543900253   
1            49.574949      118.28022162236736   
2            49.574949         81.699521264648   
3            49.574949       199.4580743910713   
4            49.574949      41.420153086217326   

                  Payment_Behaviour     Monthly_Balance Credit_Score  
0   High_spent_Small_value_payments  312.49408867943663         Good  
1    Low_spent_Large_value_payments  284.62916249607184         Good  
2   Low_spent_Medium_value_payments   331.2098628537912         Good  
3    Low_spent_Small_value_payments  223.45130972736786         Good  
4  High_spent_Medium_value_payments  341.48923103222177         Good  

[5 rows x 23 columns]

In [74]:
# Convert Age to numeric
df["Age"] = pd.to_numeric(df["Age"], errors="coerce")

In [80]:
# Fix wrong age values
df["Age"] = df["Age"].apply(lambda x: np.nan if (pd.isna(x) or x < 0 or x > 100) else x)

In [76]:
# Replace wrong values in Credit Mix
df["Credit_Mix"] = df["Credit_Mix"].replace("_", "Unknown")

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 23 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Month                     100000 non-null  object 
 1   Age                       92370 non-null   float64
 2   Occupation                100000 non-null  object 
 3   Annual_Income             100000 non-null  object 
 4   Monthly_Inhand_Salary     84998 non-null   float64
 5   Num_Bank_Accounts         100000 non-null  int64  
 6   Num_Credit_Card           100000 non-null  int64  
 7   Interest_Rate             100000 non-null  int64  
 8   Num_of_Loan               100000 non-null  object 
 9   Delay_from_due_date       100000 non-null  int64  
 10  Num_of_Delayed_Payment    92998 non-null   object 
 11  Changed_Credit_Limit      100000 non-null  object 
 12  Num_Credit_Inquiries      98035 non-null   float64
 13  Credit_Mix                100000 non-null  ob

In [46]:
df.dtypes

Month                        object
Age                         float64
Occupation                   object
Annual_Income                object
Monthly_Inhand_Salary       float64
Num_Bank_Accounts             int64
Num_Credit_Card               int64
Interest_Rate                 int64
Num_of_Loan                  object
Delay_from_due_date           int64
Num_of_Delayed_Payment       object
Changed_Credit_Limit         object
Num_Credit_Inquiries        float64
Credit_Mix                   object
Outstanding_Debt             object
Credit_Utilization_Ratio    float64
Credit_History_Age           object
Payment_of_Min_Amount        object
Total_EMI_per_month         float64
Amount_invested_monthly      object
Payment_Behaviour            object
Monthly_Balance              object
Credit_Score                 object
dtype: object

In [47]:
cols_to_convert = [
    "Num_of_Loan",
    "Num_of_Delayed_Payment",
    "Changed_Credit_Limit",
    "Amount_invested_monthly"
]

for col in cols_to_convert:
    df[col] = pd.to_numeric(df[col], errors="coerce")

In [48]:
for col in cols_to_convert:
    df[col] = df[col].fillna(0).astype(int)

In [49]:
df.dtypes

Month                        object
Age                         float64
Occupation                   object
Annual_Income                object
Monthly_Inhand_Salary       float64
Num_Bank_Accounts             int64
Num_Credit_Card               int64
Interest_Rate                 int64
Num_of_Loan                   int64
Delay_from_due_date           int64
Num_of_Delayed_Payment        int64
Changed_Credit_Limit          int64
Num_Credit_Inquiries        float64
Credit_Mix                   object
Outstanding_Debt             object
Credit_Utilization_Ratio    float64
Credit_History_Age           object
Payment_of_Min_Amount        object
Total_EMI_per_month         float64
Amount_invested_monthly       int64
Payment_Behaviour            object
Monthly_Balance              object
Credit_Score                 object
dtype: object

# **STEP 4: SPLIT X & y**

In [50]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer

In [82]:
X = df.drop("Credit_Score", axis=1)
y = df["Credit_Score"]

In [52]:
X

Month   Age Occupation Annual_Income  Monthly_Inhand_Salary  \
0       January  23.0  Scientist      19114.12            1824.843333   
1      February  23.0  Scientist      19114.12                    NaN   
2         March   NaN  Scientist      19114.12                    NaN   
3         April  23.0  Scientist      19114.12                    NaN   
4           May  23.0  Scientist      19114.12            1824.843333   
...         ...   ...        ...           ...                    ...   
99995     April  25.0   Mechanic      39628.99            3359.415833   
99996       May  25.0   Mechanic      39628.99            3359.415833   
99997      June  25.0   Mechanic      39628.99            3359.415833   
99998      July  25.0   Mechanic      39628.99            3359.415833   
99999    August  25.0   Mechanic     39628.99_            3359.415833   

       Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  \
0                      3                4              3            4   
1                      3                4              3            4   
2                      3                4              3            4   
3                      3                4              3            4   
4                      3                4              3            4   
...                  ...              ...            ...          ...   
99995                  4                6              7            2   
99996                  4                6              7            2   
99997                  4                6           5729            2   
99998                  4                6              7            2   
99999                  4                6              7            2   

       Delay_from_due_date  ...  Num_Credit_Inquiries  Credit_Mix  \
0                        3  ...                   4.0     Unknown   
1                       -1  ...                   4.0        Good   
2                        3  ...                   4.0        Good   
3                        5  ...                   4.0        Good   
4                        6  ...                   4.0        Good   
...                    ...  ...                   ...         ...   
99995                   23  ...                   3.0     Unknown   
99996                   18  ...                   3.0     Unknown   
99997                   27  ...                   3.0        Good   
99998                   20  ...                   3.0        Good   
99999                   18  ...                   3.0        Good   

       Outstanding_Debt Credit_Utilization_Ratio      Credit_History_Age  \
0                809.98                26.822620   22 Years and 1 Months   
1                809.98                31.944960                     NaN   
2                809.98                28.609352   22 Years and 3 Months   
3                809.98                31.377862   22 Years and 4 Months   
4                809.98                24.797347   22 Years and 5 Months   
...                 ...                      ...                     ...   
99995            502.38                34.663572   31 Years and 6 Months   
99996            502.38                40.565631   31 Years and 7 Months   
99997            502.38                41.255522   31 Years and 8 Months   
99998            502.38                33.638208   31 Years and 9 Months   
99999            502.38                34.192463  31 Years and 10 Months   

       Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
0                         No           49.574949                      80   
1                         No           49.574949                     118   
2                         No           49.574949                      81   
3                         No           49.574949                     199   
4                         No           49.574949                      41   
...                      ...                 ...          

In [53]:
y

0            Good
1            Good
2            Good
3            Good
4            Good
           ...   
99995        Poor
99996        Poor
99997        Poor
99998    Standard
99999        Poor
Name: Credit_Score, Length: 100000, dtype: object

In [54]:
# Categorical Columns
cat_cols = X.select_dtypes(include=["object"]).columns
num_cols = X.select_dtypes(include=["int64", "float64"]).columns

In [83]:
# Convert ALL categorical to string SAFELY
X[cat_cols] = X[cat_cols].astype(str)

In [84]:
# Target ko bhi ek hi type banao
y = y.astype(str)

# **STEP 5: PREPROCESSING**

In [55]:
numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median"))
])

In [56]:
numeric_pipeline

Pipeline(steps=[('imputer', SimpleImputer(strategy='median'))])

In [57]:
categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

In [58]:
categorical_pipeline

Pipeline(steps=[('imputer', SimpleImputer(strategy='most_frequent')),
                ('encoder', OneHotEncoder(handle_unknown='ignore'))])

In [59]:
preprocessor = ColumnTransformer([
    ("num", numeric_pipeline, num_cols),
    ("cat", categorical_pipeline, cat_cols)
])

In [60]:
preprocessor

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='median'))]),
                                 Index(['Age', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts', 'Num_Credit_Card',
       'Interest_Rate', 'Num_of_Loan', 'Delay_from_due_date',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Utilization_Ratio',
       'Total_EMI_per_month', 'Amount_invested_monthly'],
      dtype='object')),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('encoder',
                                                  OneHotEncoder(handle_unknown='ignore'))]),
                                 Index(['Month', 'Occupation', 'Annual_Income', 'Credit_Mix',
       'Outstanding_Debt', 'Credit_History_Age', 'Payment_of_Min_Amount',
       'Payment_Behaviour', 'Monthly_Balance'],
      dtype='object'))])

# **STEP 6: RANDOM FOREST MODEL**

In [61]:
rf = RandomForestClassifier(random_state=42)

In [62]:
pipeline = Pipeline([
    ("preprocess", preprocessor),
    ("rf", rf)
])

In [63]:
pipeline

Pipeline(steps=[('preprocess',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median'))]),
                                                  Index(['Age', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts', 'Num_Credit_Card',
       'Interest_Rate', 'Num_of_Loan', 'Delay_from_due_date',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Utilization_...
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  Index(['Month', 'Occupation', 'Annual_Income', 'Credit_Mix',
       'Outstanding_Debt', 'Credit_History_Age', 'Payment_of_Min_Amount',
       'Payment_Behaviour', 'Monthly_Balance'],
      dtype='object'))])),
                ('rf', RandomForestClassifier(random_state=42))])

# **STEP 7: TRAIN TEST SPLIT**

In [86]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

In [65]:
X_test

Month   Age     Occupation Annual_Income  Monthly_Inhand_Salary  \
1923      April  40.0       Mechanic      44386.47                    NaN   
17639    August  34.0      Developer      19506.73            1491.560833   
35725      June  39.0   Entrepreneur     39962.38_            3151.198333   
65889  February  28.0       Mechanic      7477.175             713.097917   
77544   January  44.0      Architect      41771.02            3293.918333   
...         ...   ...            ...           ...                    ...   
16200   January  34.0      Scientist      76338.68            6608.556667   
31780       May  38.0      Architect      32009.81            2555.484167   
28004       May  18.0  Media_Manager     107427.99            8868.332500   
31427     April  27.0      Architect      39255.54            3546.295000   
2242      March  33.0        Teacher      15713.83            1313.485833   

       Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  \
1923                  10                6             26            9   
17639                  9                7             22            6   
35725                  6                7             22            5   
65889                  9                5             15            8   
77544                  6               10             17            7   
...                  ...              ...            ...          ...   
16200                  4                6             19            1   
31780                  5             1468              9            0   
28004                  8                3             20            1   
31427                  4                3              6            2   
2242                   7                8             23            2   

       Delay_from_due_date  ...  Num_Credit_Inquiries  Credit_Mix  \
1923                    34  ...                  12.0         Bad   
17639                   47  ...                  15.0         Bad   
35725                   18  ...                   9.0     Unknown   
65889                   55  ...                   7.0     Unknown   
77544                   35  ...                  11.0         Bad   
...                    ...  ...                   ...         ...   
16200                   11  ...                   6.0     Unknown   
31780                    5  ...                   7.0     Unknown   
28004                   25  ...                   4.0     Unknown   
31427                   10  ...                   5.0        Good   
2242                    46  ...                   6.0     Unknown   

       Outstanding_Debt Credit_Utilization_Ratio      Credit_History_Age  \
1923            1275.46                25.537989    9 Years and 2 Months   
17639           1594.83                24.274661  14 Years and 10 Months   
35725           1444.91                38.856838                     NaN   
65889           4944.58                36.125353    7 Years and 4 Months   
77544           3524.73                24.735922    8 Years and 2 Months   
...                 ...                      ...                     ...   
16200            965.06                29.246919   13 Years and 3 Months   
31780            310.76                34.525378   19 Years and 3 Months   
28004           1124.02                28.410979  31 Years and 11 Months   
31427            536.03                35.147519   33 Years and 4 Months   
2242            2271.53                41.127045   10 Years and 7 Months   

       Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
1923                     Yes          298.848530                      50   
17639                    Yes           59.134042                     149   
35725                    Yes        52463.000000                     147   
65889                     NM           35.180443                       0   
77544                    Yes          127.617609                     109   
...       

In [66]:
X_train

Month   Age    Occupation Annual_Income  Monthly_Inhand_Salary  \
10268       May  51.0      Musician     101583.48                    NaN   
12727    August  23.0        Writer     101926.95            8635.912500   
30953  February  49.0        Writer     158871.12                    NaN   
61394     March  40.0        Doctor      60379.28                    NaN   
90061      June  17.0    Accountant      50050.83            4085.902500   
...         ...   ...           ...           ...                    ...   
54631    August  55.0        Doctor     114597.04                    NaN   
6534       July  28.0  Entrepreneur     8227.855_             656.654583   
21767    August  46.0        Writer      35032.66            2853.388333   
16212       May  42.0      Mechanic     129680.28           10643.690000   
92972       May  53.0     Architect     112853.22            9552.435000   

       Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  \
10268                  5                7             10            0   
12727                  4                4              9            1   
30953                  0                4              8            1   
61394                  5                6             18            3   
90061                  9               10             20            5   
...                  ...              ...            ...          ...   
54631                  7                6              4            4   
6534                   6                8             31         -100   
21767                 10                6             33            7   
16212                  8                3              5            2   
92972                  2                4              4            2   

       Delay_from_due_date  ...  Num_Credit_Inquiries  Credit_Mix  \
10268                    8  ...                   5.0    Standard   
12727                   13  ...                   6.0     Unknown   
30953                    8  ...                   4.0        Good   
61394                   15  ...                   1.0    Standard   
90061                   28  ...                   8.0         Bad   
...                    ...  ...                   ...         ...   
54631                   32  ...                   2.0        Good   
6534                    61  ...                   7.0     Unknown   
21767                   48  ...                  10.0         Bad   
16212                   27  ...                   6.0     Unknown   
92972                    3  ...                   0.0        Good   

       Outstanding_Debt Credit_Utilization_Ratio     Credit_History_Age  \
10268             50.93                34.462154  24 Years and 1 Months   
12727            1058.0                39.693812  20 Years and 5 Months   
30953            576.48                39.367225  19 Years and 0 Months   
61394            725.39                29.061701  17 Years and 1 Months   
90061            3419.1                30.386321   4 Years and 6 Months   
...                 ...                      ...                    ...   
54631            926.18                26.436313  31 Years and 9 Months   
6534            2695.38                24.127401  11 Years and 8 Months   
21767            1789.0                25.086176  11 Years and 5 Months   
16212            240.27                33.944094  20 Years and 5 Months   
92972            732.33                32.195044  17 Years and 2 Months   

       Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
10268                     No          190.811017                     630   
12727                     No           70.587681                     662   
30953                     No           86.905860                     746   
61394                     NM           90.906385                     166   
90061                    Yes          190.445060                      56   
...                      ...      

In [67]:
y_test

1923         Poor
17639        Poor
35725        Poor
65889    Standard
77544    Standard
           ...   
16200    Standard
31780    Standard
28004    Standard
31427        Good
2242         Poor
Name: Credit_Score, Length: 20000, dtype: object

In [68]:
y_train

10268    Standard
12727    Standard
30953    Standard
61394    Standard
90061        Poor
           ...   
54631        Poor
6534         Poor
21767        Poor
16212    Standard
92972    Standard
Name: Credit_Score, Length: 80000, dtype: object

In [87]:
categorical_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

# **STEP 8: HYPERPARAMETER TUNING**

In [69]:
param_grid = {
    "rf__n_estimators": [100, 200],
    "rf__max_depth": [8, 10, 12],
    "rf__min_samples_split": [2, 5],
    "rf__min_samples_leaf": [1, 2]
}

In [70]:
param_grid

{'rf__n_estimators': [100, 200],
 'rf__max_depth': [8, 10, 12],
 'rf__min_samples_split': [2, 5],
 'rf__min_samples_leaf': [1, 2]}

In [71]:
grid = GridSearchCV(
    estimator=pipeline,
    param_grid=param_grid,
    cv=3,
    scoring="accuracy",
    n_jobs=-1
)

In [72]:
grid

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('preprocess',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer(strategy='median'))]),
                                                                         Index(['Age', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts', 'Num_Credit_Card',
       'Interest_Rate', 'Num_of_Loan', 'Delay_from_due_date',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inq...
                                                                         Index(['Month', 'Occupation', 'Annual_Income', 'Credit_Mix',
       'Outstanding_Debt', 'Credit_History_Age', 'Payment_of_Min_Amount',
       'Payment_Behaviour', 'Monthly_Balance'],
      dtype='object'))])),
                                       ('rf',
                                        RandomForestClassifier(random_state=42))]),
             n_jobs=-1,
             param_grid={'rf__max_depth': [8, 10, 12],
                         'rf__min_samples_leaf': [1, 2],
                         'rf__min_samples_split': [2, 5],
                         'rf__n_estimators': [100, 200]},
             scoring='accuracy')

In [88]:
grid.fit(X_train, y_train)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('preprocess',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer(strategy='median'))]),
                                                                         Index(['Age', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts', 'Num_Credit_Card',
       'Interest_Rate', 'Num_of_Loan', 'Delay_from_due_date',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inq...
                                                                         Index(['Month', 'Occupation', 'Annual_Income', 'Credit_Mix',
       'Outstanding_Debt', 'Credit_History_Age', 'Payment_of_Min_Amount',
       'Payment_Behaviour', 'Monthly_Balance'],
      dtype='object'))])),
                                       ('rf',
                                        RandomForestClassifier(random_state=42))]),
             n_jobs=-1,
             param_grid={'rf__max_depth': [8, 10, 12],
                         'rf__min_samples_leaf': [1, 2],
                         'rf__min_samples_split': [2, 5],
                         'rf__n_estimators': [100, 200]},
             scoring='accuracy')

In [89]:

best_model = grid.best_estimator_
print("Best Parameters:", grid.best_params_)

Best Parameters: {'rf__max_depth': 8, 'rf__min_samples_leaf': 1, 'rf__min_samples_split': 2, 'rf__n_estimators': 100}


# **STEP 9: PREDICTIONS**

In [90]:
y_pred = best_model.predict(X_test)

In [91]:
print("\nAccuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Accuracy: 0.53175

Classification Report:
               precision    recall  f1-score   support

        Good       0.00      0.00      0.00      3566
        Poor       0.00      0.00      0.00      5799
    Standard       0.53      1.00      0.69     10635

    accuracy                           0.53     20000
   macro avg       0.18      0.33      0.23     20000
weighted avg       0.28      0.53      0.37     20000


Confusion Matrix:
 [[    0     0  3566]
 [    0     0  5799]
 [    0     0 10635]]


/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


# **STEP 10: FEATURE IMPORTANCE**

In [92]:
# Get feature names after encoding
ohe = best_model.named_steps["preprocess"].named_transformers_["cat"]["encoder"]

In [93]:
ohe

OneHotEncoder(handle_unknown='ignore')

In [94]:
encoded_cat_features = ohe.get_feature_names_out(cat_cols)

In [95]:
encoded_cat_features

array(['Month_April', 'Month_August', 'Month_February', ...,
       'Monthly_Balance_999.8134716639598',
       'Monthly_Balance___-333333333333333333333333333__',
       'Monthly_Balance_nan'], dtype=object)

In [96]:
final_feature_names = np.concatenate([num_cols, encoded_cat_features])

In [97]:
final_feature_names

array(['Age', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts', ...,
       'Monthly_Balance_999.8134716639598',
       'Monthly_Balance___-333333333333333333333333333__',
       'Monthly_Balance_nan'], dtype=object)

In [98]:
importances = best_model.named_steps["rf"].feature_importances_

In [99]:
importances

array([0.00165891, 0.02763425, 0.003269  , ..., 0.        , 0.        ,
       0.00017804])

In [100]:
feature_importance_df = pd.DataFrame({
    "Feature": final_feature_names,
    "Importance": importances
}).sort_values(by="Importance", ascending=False)

In [101]:
feature_importance_df

Feature  Importance
31256           Payment_of_Min_Amount_Yes    0.034643
17859                     Credit_Mix_Good    0.033858
9                    Num_Credit_Inquiries    0.033710
4                           Interest_Rate    0.029993
1                   Monthly_Inhand_Salary    0.027634
...                                   ...         ...
37689  Monthly_Balance_199.19772534413434    0.000000
37688  Monthly_Balance_199.18680103741718    0.000000
37687   Monthly_Balance_199.1862043373673    0.000000
37686   Monthly_Balance_199.1745578616164    0.000000
37699  Monthly_Balance_199.25874888905273    0.000000

[110309 rows x 2 columns]

In [102]:
print("\nTop Important Features:\n")
print(feature_importance_df.head(15))


Top Important Features:

                                                Feature  Importance
31256                         Payment_of_Min_Amount_Yes    0.034643
17859                                   Credit_Mix_Good    0.033858
9                                  Num_Credit_Inquiries    0.033710
4                                         Interest_Rate    0.029993
1                                 Monthly_Inhand_Salary    0.027634
8                                  Changed_Credit_Limit    0.022876
5                                           Num_of_Loan    0.022652
3                                       Num_Credit_Card    0.021370
6                                   Delay_from_due_date    0.021091
31255                          Payment_of_Min_Amount_No    0.019921
17860                               Credit_Mix_Standard    0.018055
12                              Amount_invested_monthly    0.017609
11                                  Total_EMI_per_month    0.016493
31263  Payment_Behavio

The model achieved 0.53175% accuracy. The most important features affecting credit score are Outstanding Debt, Interest Rate, Credit Mix, Delayed Payments and Credit Inquiries. This means credit score is highly influenced by repayment behaviour and outstanding liabilities rather than just income.